### ISSUE TITLE
UdpSocket: unnecessary mut?

### ISSUE THREAD

**Issue by sopium** [tokio-rs/tokio#615]
## Versions

tokio-udp 0.1.2

## Description

Currently, `UdpSocket` takes `&mut self` in `poll_send`, `poll_recv`, `poll_send_to` and `poll_recv_from`, and makes it unpleasant to share `UdpSocket`s.

Would it make sense to change them to take `&self`? I have tried and it compiles and works just fine. It would also align better with `AsyncRead` and `AsyncWrite` implementations for `&TcpStream` and `&UnixStream`.

**Comment by carllerche** [comment]
Tokio's types require `&mut` to ensure that they are not called (concurrently) from multiple tasks.

When `poll_recv` returns `NotReady`, the current task is registered for a wakeup notification. When poll_recv` is called a second time from a different task and returns `NotReady` again, the original task registration is erased in favor of the second task.

If `poll_recv` took `&mut self` task registration wouldn't make sense.

**Comment by sopium** [comment]
I don't think taking `&mut` ensures that:

```rust
struct RecvFromAsync<T> {
    socket: Arc<Mutex<UdpSocket>>,
    buf: T,
}

impl<T: AsMut<[u8]>> Future for RecvFromAsync<T> {
    type Item = (usize, SocketAddr);
    type Error = io::Error;

    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
        let mut socket = self.socket.lock().unwrap();
        socket.poll_recv_from(self.buf.as_mut())
    }
}
```

Multiple copies of this future can be polled from multiple tasks, and indeed there will be lost wake ups. The current API can't stop that.

And most other types does not actually require `&mut`, esp. `UnixDgram`: it takes `&self` in every poll functions.

**Comment by carllerche** [comment]
Yes, it doesn't 100% guarantee it, but it adds friction and hopefully one would stop and ask why that friction is necessary.

**Comment by sopium** [comment]
> Yes, it doesn't 100% guarantee it, but it adds friction and hopefully one would stop and ask why that friction is necessary.

Well it worked for me this time 🤔.

Still I don't think requiring `&mut` is a good idea. It should certainly be better documented. Maybe a runtime check can be implemented. But requiring `&mut` adds a lot of friction and runtime cost (e.g., `Mutex`s), and does not really achieve its goal here.

**Comment by tobz** [comment]
I don't believe that finding a way around the `&mut self` constraint is a strong enough argument for changing the signature.  The goal of requiring `&mut self` is to provide enough friction so that people either realize they're doing something unintended or ask for help and then can be told what they're doing is wrong, as @carllerche stated.

I do agree that we could document/explain this API decision better (or at all), either inline or potentially within the Tokio overview/architecture that Carl is working on currently.  It's certainly on us to be able to articulate why things are the way they are.

Do you have any thoughts on how we could both provide guardrails for normal users and allow the behavior that you're looking for in your own usage?  I see that you added some documentation notes in your PR, but, again, we want to put users on the right path by default, including the compiler enforcing this for us.

**Comment by sopium** [comment]
> The goal of requiring &mut self is to provide enough friction so that people either realize they're doing something unintended or ask for help and then can be told what they're doing is wrong, as @carllerche stated.

All this just sounds like a poor way of communication and documentation. I think many people will just wrap it in a `Mutex`/`RefCell`. After all, that is the standard way to share access to some mutable structure:

```rust
// In many different tasks. socket is a Arc<Mutex<UdpSocket>>.
socket.lock().unwrap().poll_send_to(...)
```

(I use `poll_send_to` here because it is very likely that many tasks have datagrams to send through a same socket. Heck, it's what I have written before learned about this concurrency problem. And this code works most of the time, because `sendto` rarely blocks...)

This code pays the runtime cost of `Mutex`s for nothing, and still risks lost notification and hanging.

> Do you have any thoughts on how we could both provide guardrails for normal users and allow the behavior that you're looking for in your own usage? I see that you added some documentation notes in your PR, but, again, we want to put users on the right path by default, including the compiler enforcing this for us.

I think it's pretty hard, if at all possible, to enforce this at compile time. Maybe it should be solved at run time instead? Some thoughts:

* Runtime checks. Warn that there are lost registrations. Warn that task is hanging.

* Change the reactor to allow multiple tasks waiting for a same event.

* Should there be a task equivalent of `Mutex`, it can used to enforce this rule.

**Comment by sopium** [comment]
> So, we can agree to disagree about the what the current API is accomplishing. I'd like to point out something I noticed: is there a particular reason you're trying to send to the same socket from multiple tasks?

Maybe because it's totally legitimate to send via the same socket from multiple OS threads or other languages' “task”s, and one would intuitively assume this also works in tokio.

> Normally a person would use something like an MPSC channel, where callers would send their messages into the channel, and a task receiving at the other end would be writing them to the socket.
>
> I'm curious to hear about your use case and whether or not this pattern could work for you.

Yes, that's exactly how I write it now.

**Comment by tobz** [comment]
> Maybe because it's totally legitimate to send via the same socket from multiple OS threads or other languages' “task”s, and one would intuitively assume this also works in tokio.

Agreed, and I think the documentation that Carl is working on will help explain why this operation, which works in other languages/frameworks, doesn't work in Tokio.

I'll make sure we track this issue as we work on the documentation and hopefully we can come up with something suitable.

**Comment by carllerche** [comment]
Additionally, this change would prohibit an optimization that has yet to be made: avoiding an atomic call to get the readiness value cache.

**Comment by 0xcaff** [comment]
FWIW, I've gotten around this by creating a `std::net::UdpSocket`, cloning it, and using `UdpSocket::from_std`. I use one `tokio::net::UdpSocket` for sending and one for receiving.
